File: 2020/day07/haskell/day72.hs
import Data.Map (fromList, (!), Map)

main :: IO()
main = interact $ show 
    . (\bagMap -> 
        sum 
        $ map 
            (\tup -> search (fst tup) (snd tup) bagMap) 
            (bagMap ! "shinygold")
      )
    . fromList . map (parse . words) . lines
  
search :: Int -> String -> Map String [(Int, String)] -> Int
search n name di
    | name == "empty" = 0
    | otherwise = (+n) $ (*n) $ sum 
        $ map (\tup -> search (fst tup) (snd tup) di) (di ! name)

parse :: [String] -> (String, [(Int, String)])
parse bag = (concat (take 2 bag), splitBags (drop 4 bag))

splitBags :: [String] -> [(Int, String)]
splitBags [] = []
splitBags (x:y:z:_:xs) = (read x, y ++ z):splitBags xs
splitBags _ = [(1, "empty")]

File: 2020/day07/python/day72.py
def parse(line):
    words = line.split()
    name = "".join(words[:2])
    elems = split_bags(words[4:])
    return (name, elems)


def divide_chunks(l, n):
    for i in range(0, len(l), n):
        yield l[i:i + n]


def split_bags(word_list):
    return [(bag[0], bag[1] + bag[2])
            if bag[0] != "no"
            else (0, "empty")
            for bag in divide_chunks(word_list, 4)]


def search(n, name, di):
    if name == "empty": return 0
    return n + sum(map(lambda tup: search(int(tup[0]), tup[1], di), di[name])) * n


def main():
    di = dict( [parse(line) for line in open("input.txt")])
    di["empty"] = (0, "")
    res = sum(map(lambda tup: search(int(tup[0]), tup[1], di), di["shinygold"]))
    print(res)


if _name_ == "_main_":
    main()